estimagic/decorators.py

```diff
@@ -0,0 +1,140 @@
+import functools
+import itertools
+
+import numpy as np
+
+from estimagic.config import DEFAULT_CRITERION_PENALTY
+from estimagic.logging.update_database import append_rows
+from estimagic.logging.update_database import update_scalar_field
+from estimagic.optimization.reparametrize import reparametrize_from_internal
+
+
+def x_to_params(params, constraints=None):
```

Comment:
I think this decorator should get one additional optional argument called `output_type` or `out_type`. If this is `pd.Series` or `pd.DataFrame` the result is converted to an array. 

Also, how about calling it `numpy_interface`? I think it looks intuitive to read something like:
```python
@numpy_interface(params, constraints, out_type=pd.Series)
def my_pandas_criterion_function(params):
    pass
```
This would also make it clearer that the decorator is something that operates on a function.

Reply:
I am not sure whether out_type is really necessary.

1. ``out_type`` is not very intuitive as I would expect that the return after decorating is a pandas series.
2. numpy interface suggests by its name that the return is a numpy array, thus it is clear to which object we want to transform.
3. Every ``out_type`` needs its own conversion to NumPy anyway. So no need to pass it in. We can just check inside the function.

But, probably you have a use-case in mind which I do not.

Reply:
No, you are right. Checking the type is better than the argument. It is just important that the output is converted to numpy if it is a pandas object and If I'm correct this is currently not done.